### DISCUSSION TITLE
Displaying Polars Dataframe

### DISCUSSION THREAD

**Discussion by frankvgompel** [emilk/egui#2222] [Q&A]
What is the simple way to display a Polars Dataframe? 

[https://pola-rs.github.io/polars/polars/index.html](url)

It displays properly in the console so it should easily be supported?
This is how a search result looks in egui with: 

`ui.label(format!("Result {:?}", result_df));`

<img width="713" alt="example_egui" src="https://user-images.githubusercontent.com/81262671/199923065-c91c291a-5873-4b15-bd3d-a493ef4391a5.png">

And this is how it prints in the console:
<img width="521" alt="example_console" src="https://user-images.githubusercontent.com/81262671/199923155-d2310467-dffb-4423-92ba-261e9439546e.png">

**Comment by fedecarles** [comment]
For anybody interested, you can display polars DataFrames using the egui_extras TableBuilder

Example, where self.data is a DataFrame. Its fairly performant, it loads dataframes of 700k+ rows without issues.

```
window.show(ctx, |ui| {
    let nr_cols = self.data.width();
    let nr_rows = self.data.height();
    let cols = &self.data.get_column_names();

    TableBuilder::new(ui)
        .columns(Column::auto(), nr_cols)
        .striped(true)
        .resizable(true)
        .header(5.0, |mut header| {
            for head in cols {
                header.col(|ui| {
                    ui.heading(format!("{}", head));
                });
            }
        })
        .body(|body| {
            body.rows(10.0, nr_rows, |row_index, mut row| {
                for col in cols {
                    row.col(|ui| {
                        if let Ok(column) = &self.data.column(col) {
                            if let Ok(value) = column.get(row_index) {
                                ui.label(format!("{}", value));
                            }
                        }
                    });
                }
            });
        });
    });
```

![egui_dataframe](https://github.com/emilk/egui/assets/1060450/cd3a962b-b899-458e-a8a8-33f4d927618d)

**Comment by YgorSouza** [comment]
These border characters and spacing only work in monospace fonts, so if you really want to display a dataframe as a string, you can just tell egui use a monospace font. You can also use a `TextEdit` if you want to be able to select and copy text (but it's going to copy the borders and spaces as well, of course).

```rs
impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        egui::CentralPanel::default().show(ctx, |ui| {
            let mut data_str = r#"
┌──────────┬──────────┬──────────────┬─────┬─────────────┬─────────────┬─────────────┬─────────────┐
│ fruits   ┆ cars     ┆ literal_stri ┆ B   ┆ sum_A_by_ca ┆ sum_A_by_fr ┆ rev_A_by_fr ┆ sort_A_by_B │
│ ---      ┆ ---      ┆ ng_fruits    ┆ --- ┆ rs          ┆ uits        ┆ uits        ┆ _by_fruits  │
│ str      ┆ str      ┆ ---          ┆ i64 ┆ ---         ┆ ---         ┆ ---         ┆ ---         │
│          ┆          ┆ str          ┆     ┆ i64         ┆ i64         ┆ i64         ┆ i64         │
╞══════════╪══════════╪══════════════╪═════╪═════════════╪═════════════╪═════════════╪═════════════╡
│ "apple"  ┆ "beetle" ┆ "fruits"     ┆ 11  ┆ 4           ┆ 7           ┆ 4           ┆ 4           │
│ "apple"  ┆ "beetle" ┆ "fruits"     ┆ 11  ┆ 4           ┆ 7           ┆ 3           ┆ 3           │
│ "banana" ┆ "beetle" ┆ "fruits"     ┆ 11  ┆ 4           ┆ 8           ┆ 5           ┆ 5           │
│ "banana" ┆ "audi"   ┆ "fruits"     ┆ 11  ┆ 2           ┆ 8           ┆ 2           ┆ 2           │
│ "banana" ┆ "beetle" ┆ "fruits"     ┆ 11  ┆ 4           ┆ 8           ┆ 1           ┆ 1           │
└──────────┴──────────┴──────────────┴─────┴─────────────┴─────────────┴─────────────┴─────────────┘
                "#;
            ui.monospace(data_str);
            ui.add(
                TextEdit::multiline(&mut data_str)
                    .desired_width(ui.available_width())
                    .code_editor(),
            );
        });
    }
}
```

![image](https://github.com/emilk/egui/assets/43298013/a9fc5f59-4c23-4fac-a97d-ab1fbc7db914)

Copied text from the selection in the image:
```
"apple"  ┆ "beetle" ┆ "fruits"     ┆ 11  ┆ 4           ┆ 7           ┆ 4           ┆ 4          
```

But in most cases you probably want to use a `Table` as the comment above said. Text rendering doesn't scale very well, as discussed in issues like #3086, and the table takes care of that by only rendering the part that is visible in each frame.
